contracts/token/ERC20WithPermit.sol

```diff
@@ -1,7 +1,314 @@
-contract Erc20WithPermit {
-    // TODO will be implemented separately
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./IERC20WithPermit.sol";
+import "./IReceiveApproval.sol";
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+contract ERC20WithPermit is IERC20WithPermit, Ownable {
+    /// @notice The amount of tokens owned by the given account.
+    mapping(address => uint256) public override balanceOf;
+
+    /// @notice The remaining number of tokens that spender will be
+    ///         allowed to spend on behalf of owner through `transferFrom` and
+    ///         `burnFrom`. This is zero by default.
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    /// @notice Returns the current nonce for EIP2612 permission for the
+    ///         provided token owner for a replay protection. Used to construct
+    ///         EIP2612 signature provided to `permit` function.
+    mapping(address => uint256) public override nonces;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    /// @notice Returns EIP2612 Permit message hash. Used to construct EIP2612
+    ///         signature provided to `permit` function.
+    // keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 public constant override PERMIT_TYPEHASH =
+        0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+
+    /// @notice The amount of tokens in existence.
+    uint256 public override totalSupply;
+
+    /// @notice The name of the token.
+    string public override name;
+
+    /// @notice The symbol of the token.
+    string public override symbol;
+
+    /// @notice The decimals places of the token.
+    uint8 public constant override decimals = 18;
+
+    constructor(string memory _name, string memory _symbol) {
+        name = _name;
+        symbol = _symbol;
+
+        cachedChainId = chainId();
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    /// @notice Moves `amount` tokens from the caller's account to `recipient`.
+    /// @return True if the operation succeeded, reverts otherwise.
+    /// @dev Requirements:
+    ///       - `recipient` cannot be the zero address,
+    ///       - the caller must have a balance of at least `amount`.
+    function transfer(address recipient, uint256 amount)
+        external
+        override
+        returns (bool)
+    {
+        _transfer(msg.sender, recipient, amount);
```

Comment:
Shouldn't we use `transferFrom` here? Aren't we missing allowance check right now?